### PR TITLE
fix: add block rewards to `trace_block`

### LINKED ISSUE #3443: Add block rewards to `trace_block` response
### Describe the feature

the `trace_block` endpoint should return the block rewards:

ref https://github.com/paradigmxyz/reth/issues/3427#issuecomment-1609902468

## TODO
* update rpc-types
* add block rewards

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by onbjerg**
Closes #3443

### CODE REVIEW

File: crates/rpc/rpc/src/trace.rs [hunk 1]
```diff
@@ -408,6 +409,42 @@ where
             )
             .await?
             .map(|traces| traces.into_iter().flatten().collect());
+
+        // Add block reward traces
+        // TODO: We only really need the header and ommers here to determine the reward
```
Comment (mattsse) [inline crates/rpc/rpc/src/trace.rs:414]: I guess we could also use the `CfgEnv`'s specId to determine the blockreward

File: crates/rpc/rpc/src/trace.rs [hunk 2]
```diff
@@ -408,6 +411,43 @@ where
...
+        if let (Some(block), Some(traces)) =
+            (self.inner.eth_api.block_by_id(block_id).await?, traces.as_mut())
+        {
+            if let Some(header_td) = self.provider().header_td(&block.header.hash)? {
+                if let Some(base_block_reward) = base_block_reward(
```
Comment (onbjerg) [inline crates/rpc/rpc/src/trace.rs:421]: if this is none we are past the merge and there are no block rewards and no ommers

File: crates/storage/provider/src/traits/spec.rs [hunk 3]
```diff
@@ -0,0 +1,8 @@
+use reth_primitives::ChainSpec;
+use std::sync::Arc;
+
+/// A trait for reading the current chainspec.
+pub trait ChainSpecReader: Send + Sync {
```
Comment (mattsse) [inline crates/storage/provider/src/traits/spec.rs:5]:
```suggestion
pub trait ChainSpecProvider: Send + Sync {
```

File: crates/storage/provider/src/traits/spec.rs [hunk 4]
```diff
@@ -0,0 +1,8 @@
...
+    /// Get an [`Arc`] to the chainspec.
+    fn spec(&self) -> Arc<ChainSpec>;
```
Comment (mattsse) [inline crates/storage/provider/src/traits/spec.rs:7]:
```suggestion
    fn chain_spec(&self) -> Arc<ChainSpec>;
```

File: crates/rpc/rpc/src/trace.rs [hunk 5]
```diff
@@ -396,7 +399,7 @@ where
         &self,
         block_id: BlockId,
     ) -> EthResult<Option<Vec<LocalizedTransactionTrace>>> {
-        let traces = self
+        let mut traces: Option<Vec<LocalizedTransactionTrace>> = self
```
Comment (mattsse) [inline crates/rpc/rpc/src/trace.rs:402]: perf: can use `futures::try_join` with the cache call

**Review by mattsse — CHANGES_REQUESTED** [review]
naming nits,

`Reader` doesn't make sense here imo, so we should stick with `Provider` naming

File: crates/rpc/rpc/src/trace.rs (hunk 5 above)
Comment (gakonst) [reply crates/rpc/rpc/src/trace.rs:402]: let's do in follow-up as i'd like to unblock this asap

File: crates/rpc/rpc/src/eth/api/transactions.rs [hunk 6]
```diff
@@ -218,6 +218,7 @@ pub trait EthTransactions: Send + Sync {
             + Send;
 }
 
+// todo: add header td call
```
Comment (gakonst) [inline crates/rpc/rpc/src/eth/api/transactions.rs]: im removing this todo - not sure what it's about, we can fix in followup if it's relevant still

File: crates/rpc/rpc/src/trace.rs [hunk 7]
```diff
@@ -408,6 +411,43 @@ where
...
+                                author: block.header.beneficiary,
+                                reward_type: RewardType::Uncle,
+                                value: block_reward(base_block_reward, block.ommers.len()) -
+                                    U256::from(base_block_reward),
+                            },
+                        ));
+                    }
+                }
+            }
+        }
```
Comment (gakonst) [inline crates/rpc/rpc/src/trace.rs:449]: nice

**Review by gakonst — APPROVED** [review]

**Review by mattsse — APPROVED** [review]

**Pull request merged**